src/ipc/ipc4/helper.c

```diff
@@ -453,32 +459,88 @@ int ipc_comp_connect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
+					  sizeof(*src_basecfg_ext) + size);
+		if (!src_basecfg_ext)
+			return IPC4_OUT_OF_MEMORY;
+
+		ret = comp_get_attribute(source, COMP_ATTR_BASE_CONFIG_EXT, src_basecfg_ext);
+		if (!ret) {
+			size_t in_fmt_size = sizeof(struct ipc4_input_pin_format);
+			size_t out_fmt_size = sizeof(struct ipc4_output_pin_format);
+
+			/*
+			 * offset of the format for the output pin with ID 'src_queue' within the
+			 * base config extension.
+			 */
+			output_fmt_offset =
+				src_basecfg_ext->nb_input_pins * in_fmt_size +
+				bu->extension.r.src_queue * out_fmt_size;
+			out_fmt = (struct ipc4_output_pin_format *)&sink_basecfg_ext->pin_formats[output_fmt_offset];
+			obs = out_fmt->obs;
+		}
+		rfree(src_basecfg_ext);
 	}
 
-	ret = comp_get_attribute(sink, COMP_ATTR_BASE_CONFIG, &sink_src_cfg);
-	if (ret < 0) {
-		tr_err(&ipc_tr, "failed to get base config for module %#x", dev_comp_id(sink));
-		return IPC4_FAILURE;
+	/* get obs from base config if src queue ID is 0 or if base config extn is missing */
+	if (!obs) {
+		/* these might call comp_ipc4_get_attribute_remote() if necessary */
+		ret = comp_get_attribute(source, COMP_ATTR_BASE_CONFIG, &source_src_cfg);
+
+		if (ret < 0) {
+			tr_err(&ipc_tr, "failed to get base config for src module %#x",
+			       dev_comp_id(source));
+			return IPC4_FAILURE;
+		}
+		obs = source_src_cfg.obs;
 	}
 
-	/* create a buffer
-	 * in case of LL -> LL or LL->DP
-	 *	size = 2*obs of source module (obs is single buffer size)
-	 * in case of DP -> LL
-	 *	size = 2*ibs of destination (LL) module. DP queue will handle obs of DP module
-	 */
-	uint32_t buf_size;
+	/* get ibs from the base config extension if the sink queue ID is non-zero */
+	if (bu->extension.r.dst_queue) {
+		struct ipc4_input_pin_format *in_fmt;
+		size_t input_fmt_offset;
+
+		sink_basecfg_ext = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+					   sizeof(*sink_basecfg_ext) + size);
+		if (!sink_basecfg_ext)
+			return IPC4_OUT_OF_MEMORY;
+
+		ret = comp_get_attribute(sink, COMP_ATTR_BASE_CONFIG_EXT, sink_basecfg_ext);
+		if (!ret) {
+			/*
+			 * offset of the format for the input pin with ID 'dst_queue' within the
+			 * base config extension.
+			 */
+			input_fmt_offset =
+				bu->extension.r.dst_queue * sizeof(struct ipc4_input_pin_format);
+			in_fmt = (struct ipc4_input_pin_format *)&sink_basecfg_ext->pin_formats[input_fmt_offset];
+			ibs = in_fmt->ibs;
+		}
+		rfree(sink_basecfg_ext);
+	}
 
-	if (source->ipc_config.proc_domain == COMP_PROCESSING_DOMAIN_LL)
-		buf_size = source_src_cfg.obs * 2;
-	else
-		buf_size = sink_src_cfg.ibs * 2;
+	/* get ibs from base config if sink queue ID is 0 or if base config extn is missing */
+	if (!ibs) {
+		ret = comp_get_attribute(sink, COMP_ATTR_BASE_CONFIG, &sink_src_cfg);
+		if (ret < 0) {
+			tr_err(&ipc_tr, "failed to get base config for sink module %#x",
+			       dev_comp_id(sink));
+			return IPC4_FAILURE;
+		}
+
+		ibs = sink_src_cfg.ibs;
+	}
```

Comment:
Another minor note: would be nice to split out into subfunctions somehow, this function is getting very, very long.